crates/net/eth-wire/src/types/message.rs

```diff
@@ -237,6 +237,124 @@ impl EthMessage {
+    /// The resulting data will fit within the limit but may be incomplete if truncation occurs.
+    ///
+    /// When using this method with collections or "vec-like" structures, implementors should take
+    /// care to ensure the truncated encoding remains valid and interpretable, albeit potentially
+    /// incomplete. For example, truncating a list of items might result in the list containing
+    /// fewer items than intended, but each item in the list should remain individually intact
+    /// and decodable.
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing the encoded data, truncated to fit within the size limit if necessary.
+    /// Users of this method should be aware that the truncated data may not represent the complete
+    /// original dataset.
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8>;
+
+    /// Encodes the data up to the specified size limit, returning the encoded data along with
+    /// a flag indicating whether the limit was exceeded.
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The encoded data as `Vec<u8>`.
+    /// * The length of the data at the point just before reaching the limit, or 0 if the limit was
+    ///   not exceeded.
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize);
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(
+        &self,
+        approx: usize,
+        limit: usize,
+    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
+        let (mut buf, current_len) = self.encode_until_limit(approx, limit);
+
+        if current_len != 0 {
+            // Don't shrink if limit_exceeded since we return Error anyway
+            return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+        }
+        buf.shrink_to_fit();
+        Ok(buf)
+    }
+
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize) {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        for item in self {
+            header.payload_length += item.length();
+        }
+        header.encode(&mut buffer);
+        for item in self {
+            let current_len = buffer.len();
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return (buffer, current_len);
+            }
+        }
```

Comment:
this includes an invalid header if items are skipped
we either need to write the header last or replace in case buffer.len() > limit

Reply:
Got it

Reply:
it should be possible to just use `.length()` to determine how many elements to write and calculate the header payload len beforehand

Reply:
that's the best solution yes

Reply:
Pushed a change, it assumes the list header is <= 3